File: editors/intellij/src/main/kotlin/com/mediusfy/modulex/intellij/core/McpClient.kt
// Minimal MCP client over the stdio transport: newline-delimited JSON-RPC
// 2.0, matching the Go SDK's mcp.StdioTransport that tools/mcpserver runs.
// Hand-rolled (like editors/vscode/src/mcpClient.ts) so the plugin needs no
// MCP SDK and works fully offline.
//
// The client is synchronous and stream-based: requests are serialized under
// a lock (the plugin issues one tool call at a time from a background
// task), and the transport is a plain Reader/Writer pair so tests drive it
// with in-memory pipes — no subprocess required.
package com.mediusfy.modulex.intellij.core

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File

class McpException(message: String) : Exception(message)

interface McpTransport {
    val reader: BufferedReader
    val writer: BufferedWriter

    fun close()
}

data class ServerSpec(val command: List<String>, val workingDirectory: File)

private class ProcessTransport(private val process: Process, onServerLog: (String) -> Unit) :
    McpTransport {
    override val reader: BufferedReader = process.inputStream.bufferedReader()
    override val writer: BufferedWriter = process.outputStream.bufferedWriter()

    init {
        Thread {
            process.errorStream.bufferedReader().forEachLine(onServerLog)
        }.apply {
            isDaemon = true
            name = "modulex-mcp-stderr"
            start()
        }
    }

    override fun close() {
        process.destroy()
    }
}

class McpClient(private val transport: McpTransport) {
    private val gson = Gson()
    private var nextId = 1
    private val lock = Any()

    companion object {
        const val PROTOCOL_VERSION = "2024-11-05"

        fun start(spec: ServerSpec, onServerLog: (String) -> Unit): McpClient {
            val process =
                try {
                    ProcessBuilder(spec.command).directory(spec.workingDirectory).start()
                } catch (e: Exception) {
                    throw McpException("modulex MCP server failed to start: ${e.message}")
                }
            val client = McpClient(ProcessTransport(process, onServerLog))
            client.initialize()
            return client
        }
    }

    fun initialize() {
        val params = JsonObject()
        params.addProperty("protocolVersion", PROTOCOL_VERSION)
        params.add("capabilities", JsonObject())
        val clientInfo = JsonObject()
        clientInfo.addProperty("name", "modulex-intellij")
        clientInfo.addProperty("version", "0.1.0")
        params.add("clientInfo", clientInfo)
        request("initialize", params)
        notify("notifications/initialized", JsonObject())
    }

    fun callTool(name: String, args: JsonObject): JsonObject {
        val params = JsonObject()
        params.addProperty("name", name)
        params.add("arguments", args)
        val result = request("tools/call", params)
        val text =
            (result.getAsJsonArray("content") ?: com.google.gson.JsonArray())
                .filterIsInstance<JsonObject>()
                .filter { it.get("type")?.asString == "text" }
                .mapNotNull { it.get("text")?.asString }
                .joinToString("\n")
        if (result.get("isError")?.asBoolean == true) {
            throw McpException(text.ifEmpty { "tool $name failed" })
        }
        result.get("structuredContent")?.let {
            if (it.isJsonObject) {
                return it.asJsonObject
            }
        }
        return try {
            JsonParser.parseString(text).asJsonObject
        } catch (_: Exception) {
            throw McpException("tool $name returned no structured content and non-JSON text: $text")
        }
    }

    private fun request(method: String, params: JsonObject): JsonObject =
        synchronized(lock) {
            val id = nextId++
            writeMessage(id, method, params)
            // Read until this request's response; the server may interleave
            // notifications (no id), which are skipped.
            while (true) {
                val line = transport.reader.readLine() ?: throw McpException(
                    "modulex MCP server closed the connection",
                )
                if (line.isBlank()) {
                    continue
                }
                val msg =
                    try {
                        JsonParser.parseString(line).asJsonObject
                    } catch (_: Exception) {
                        continue // not JSON-RPC; ignore, matching the VSCode client
                    }
                if (msg.get("id")?.takeIf { it.isJsonPrimitive }?.asInt != id) {
                    continue
                }
                msg.getAsJsonObject("error")?.let {
                    throw McpException(it.get("message")?.asString ?: "JSON-RPC error")
                }
                return msg.getAsJsonObject("result") ?: JsonObject()
            }
            @Suppress("UNREACHABLE_CODE")
            throw IllegalStateException("unreachable")
        }

    private fun notify(method: String, params: JsonObject) {
        synchronized(lock) { writeMessage(null, method, params) }
    }

    private fun writeMessage(id: Int?, method: String, params: JsonObject) {
        val msg = JsonObject()
        msg.addProperty("jsonrpc", "2.0")
        if (id != null) {
            msg.addProperty("id", id)
        }
        msg.addProperty("method", method)
        msg.add("params", params)
        transport.writer.write(gson.toJson(msg))
        transport.writer.write("\n")
        transport.writer.flush()
    }

    fun dispose() {
        transport.close()
    }
}

File: editors/vscode/src/mcpClient.ts
// Minimal MCP client over the stdio transport: newline-delimited JSON-RPC
// 2.0, matching the Go SDK's mcp.StdioTransport that tools/mcpserver runs.
// Hand-rolled instead of depending on an SDK so the extension has zero
// runtime dependencies and works fully offline.

import { spawn, type ChildProcess } from "node:child_process";
import * as readline from "node:readline";

export interface ServerSpec {
  command: string;
  args: string[];
  cwd: string;
}

interface JsonRpcResponse {
  jsonrpc: "2.0";
  id?: number;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
  method?: string;
}

interface ToolCallResult {
  isError?: boolean;
  structuredContent?: unknown;
  content?: Array<{ type: string; text?: string }>;
}

const PROTOCOL_VERSION = "2024-11-05";

export class McpError extends Error {}

export class McpClient {
  private nextId = 1;
  private readonly pending = new Map<
    number,
    { resolve: (v: unknown) => void; reject: (e: Error) => void }
  >();
  private stderrTail: string[] = [];
  private exited = false;

  private constructor(
    private readonly proc: ChildProcess,
    private readonly log: (line: string) => void,
  ) {}

  static async start(
    spec: ServerSpec,
    log: (line: string) => void,
  ): Promise<McpClient> {
    const proc = spawn(spec.command, spec.args, {
      cwd: spec.cwd,
      stdio: ["pipe", "pipe", "pipe"],
    });
    const client = new McpClient(proc, log);
    client.wire();
    await client.request("initialize", {
      protocolVersion: PROTOCOL_VERSION,
      capabilities: {},
      clientInfo: { name: "modulex-vscode", version: "0.1.0" },
    });
    client.notify("notifications/initialized", {});
    return client;
  }

  private wire(): void {
    const stdout = this.proc.stdout;
    if (stdout) {
      readline.createInterface({ input: stdout }).on("line", (line) => {
        this.onLine(line);
      });
    }
    const stderr = this.proc.stderr;
    if (stderr) {
      readline.createInterface({ input: stderr }).on("line", (line) => {
        this.stderrTail.push(line);
        if (this.stderrTail.length > 50) {
          this.stderrTail.shift();
        }
        this.log(`[server] ${line}`);
      });
    }
    this.proc.on("error", (err) => {
      this.failAll(
        new McpError(`modulex MCP server failed to start: ${err.message}`),
      );
    });
    this.proc.on("exit", (code) => {
      this.exited = true;
      const detail = this.stderrTail.length
        ? `; server said: ${this.stderrTail.slice(-5).join(" | ")}`
        : "";
      this.failAll(
        new McpError(
          `modulex MCP server exited with code ${code ?? "unknown"}${detail}`,
        ),
      );
    });
  }

  private onLine(line: string): void {
    const trimmed = line.trim();
    if (trimmed === "") {
      return;
    }
    let msg: JsonRpcResponse;
    try {
      msg = JSON.parse(trimmed) as JsonRpcResponse;
    } catch {
      this.log(`[server stdout, not JSON-RPC] ${trimmed}`);
      return;
    }
    if (typeof msg.id !== "number") {
      return; // notification from the server; nothing to route
    }
    const waiter = this.pending.get(msg.id);
    if (!waiter) {
      return;
    }
    this.pending.delete(msg.id);
    if (msg.error) {
      waiter.reject(new McpError(msg.error.message));
    } else {
      waiter.resolve(msg.result);
    }
  }

  private failAll(err: Error): void {
    for (const waiter of this.pending.values()) {
      waiter.reject(err);
    }
    this.pending.clear();
  }

  private write(msg: object): void {
    if (this.exited || !this.proc.stdin || !this.proc.stdin.writable) {
      throw new McpError("modulex MCP server is not running");
    }
    this.proc.stdin.write(`${JSON.stringify(msg)}\n`);
  }

  private request(method: string, params: object): Promise<unknown> {
    const id = this.nextId++;
    return new Promise((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
      try {
        this.write({ jsonrpc: "2.0", id, method, params });
      } catch (err) {
        this.pending.delete(id);
        reject(err instanceof Error ? err : new McpError(String(err)));
      }
    });
  }

  private notify(method: string, params: object): void {
    this.write({ jsonrpc: "2.0", method, params });
  }

  get alive(): boolean {
    return !this.exited;
  }

  async callTool<T>(name: string, args: object): Promise<T> {
    const raw = (await this.request("tools/call", {
      name,
      arguments: args,
    })) as ToolCallResult;
    const text = (raw.content ?? [])
      .filter((c) => c.type === "text" && typeof c.text === "string")
      .map((c) => c.text)
      .join("\n");
    if (raw.isError) {
      throw new McpError(text || `tool ${name} failed`);
    }
    if (raw.structuredContent !== undefined) {
      return raw.structuredContent as T;
    }
    try {
      return JSON.parse(text) as T;
    } catch {
      throw new McpError(
        `tool ${name} returned no structured content and non-JSON text: ${text}`,
      );
    }
  }

  dispose(): void {
    this.exited = true;
    this.failAll(new McpError("modulex MCP client disposed"));
    this.proc.kill();
  }
}
